server/src/test/java/org/spine3/server/procman/IteratingCommandRouterShould.java

```diff
@@ -0,0 +1,148 @@
+    private Message sourceMessage;
+
+    /** The context of the command that we route. */
+    private CommandContext sourceContext;
+
+    /** Command messages to be sent. */
+    private final List<Message> messages = ImmutableList.<Message>of(
+            newStringValue("uno"),
+            newStringValue("dos"),
+            newStringValue("tres"),
+            newStringValue("cuatro")
+    );
+
+    @Before
+    public void setUp() {
+        final BoundedContext boundedContext = BoundedContext.newBuilder()
+                                                            .build();
+        final CommandBus commandBus = boundedContext.getCommandBus();
+        commandBus.register(new CommandDispatcher() {
+            @Override
+            public Set<CommandClass> getCommandClasses() {
+                return CommandClass.setOf(StringValue.class);
+            }
+
+            @Override
+            public void dispatch(Command request) throws Exception {
+                // Do nothing.
+            }
+        });
+
+        sourceMessage = newStringValue(getClass().getSimpleName());
+        sourceContext = commandFactory.createCommandContext();
+
+        router = new IteratingCommandRouter(commandBus, sourceMessage, sourceContext);
+        router.addAll(messages);
+    }
+
+    @Test
+    public void routeFirst() throws Exception {
+        final CommandRouted commandRouted = router.routeFirst();
+
+        // Check that the source command is stored.
+        final Command source = commandRouted.getSource();
+        assertEquals(sourceMessage, Commands.getMessage(source));
+        assertEquals(sourceContext, source.getContext());
+
+        // Test that only only one command was produced by `routeFirst()`.
+        assertEquals(1, commandRouted.getProducedCount());
+
+        // Test that there's only one produced command and it has correct message.
+        final Command produced = commandRouted.getProduced(0);
+        final StringValue commandMessage = Commands.getMessage(produced);
+        assertEquals(messages.get(0), commandMessage);
+
+        // Test that the produced command context has correct fields.
+        assertTrue(Commands.sameActorAndTenant(sourceContext, produced.getContext()));
+
+        // Test that the event contains messages to follow.
+        assertEquals(messages.size() - 1, commandRouted.getMessageToFollowCount());
+
+        final List<Any> messageToFollow = commandRouted.getMessageToFollowList();
+        assertArrayEquals(messages.subList(1, messages.size()).toArray(),
+                          unpackAll(messageToFollow).toArray());
+    }
+
+    @Test
+    public void routeNext() throws Exception {
+        router.routeFirst();
+
+        final Command command = router.routeNext();
+
+        // Test that 2nd command message is in the next rounded command.
+        assertEquals(messages.get(1), Commands.getMessage(command));
+    }
+
+    private static List<StringValue> unpackAll(List<Any> anyList) {
```

Comment:
We'll on on 2nd occasion for such a need. We may want to have unpacking iterator instead so that we load from storage and un-pack on the fly without loading everything into memory. So, it's a bit more complicated.